android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFirebaseMessagingDelegate.kt

```diff
@@ -0,0 +1,89 @@
+package host.exp.exponent.fcm
+
+import android.content.Context
+import com.google.firebase.messaging.RemoteMessage
+import expo.modules.notifications.notifications.model.NotificationContent
+import expo.modules.notifications.notifications.model.NotificationRequest
+import expo.modules.notifications.notifications.model.triggers.FirebaseNotificationTrigger
+import expo.modules.notifications.service.delegates.FirebaseMessagingDelegate
+import host.exp.exponent.ABIVersion
+import host.exp.exponent.Constants
+import host.exp.exponent.analytics.EXL
+import host.exp.exponent.kernel.ExperienceId
+import host.exp.exponent.notifications.PushNotificationHelper
+import host.exp.exponent.notifications.model.ScopedNotificationRequest
+import host.exp.exponent.storage.ExponentDB
+import org.json.JSONException
+import org.json.JSONObject
+
+class ExpoFirebaseMessagingDelegate(context: Context) : FirebaseMessagingDelegate(context) {
+  override fun onNewToken(token: String) {
+    if (!Constants.FCM_ENABLED) {
+      return
+    }
+    super.onNewToken(token)
+    FcmRegistrationIntentService.registerForeground(context.applicationContext, token)
+  }
+
+  override fun onMessageReceived(remoteMessage: RemoteMessage) {
+    if (!Constants.FCM_ENABLED) {
+      return
+    }
+    val experienceDBObject = ExponentDB.experienceIdToExperienceSync(remoteMessage.data["experienceId"])
+    if (experienceDBObject != null) {
+      try {
+        val manifest = JSONObject(experienceDBObject.manifest)
+        val sdkVersion = ABIVersion.toNumber(manifest.getString("sdkVersion")) / 10000
+
+        // If an experience is on SDK newer than 39, that is SDK40 and beyond up till UNVERSIONED
+        // we only use the new notifications API as it is going to be removed from SDK40.
+        if (sdkVersion >= 40) {
+          dispatchToNextNotificationModule(remoteMessage)
+          return
+        } else if (sdkVersion == 38 || sdkVersion == 39) {
+          // In SDK38 and 39 we want to let people decide which notifications API to use,
+          // the next or the legacy one.
+          val androidSection = manifest.optJSONObject("android")
+          if (androidSection != null) {
+            val useNextNotificationsApi = androidSection.optBoolean("useNextNotificationsApi", false)
+            if (useNextNotificationsApi) {
+              dispatchToNextNotificationModule(remoteMessage)
+              return
+            }
+          }
```

Comment:
```suggestion
          androidSection?.let {
            val useNextNotificationsApi = androidSection.optBoolean("useNextNotificationsApi", false)
            if (useNextNotificationsApi) {
              dispatchToNextNotificationModule(remoteMessage)
              return@let // not sure here though 😅
            }
          }
```
more kotlin-styled 😉